MAC changes

markuskurbel opened the pull request:
An additional state responsible for the pg_hba creation would be nice as well, since the scenario where one user has multiple entries is not handled atm.

markuskurbel committed:
Added a CREATE EXTENSION state for postgis

markuskurbel committed:
Added a CREATE SCHEMA state

markuskurbel committed:
Added a state to reload the PG configs

markuskurbel committed:
Fixed some issues with the postgis state:

-Fixed indemtation of the create extension state(s)
-Renamed the create extension state
-Moved the gisdb config to the database key

markuskurbel committed:
Splitted the for loop for finer control over DB and schema creation

markuskurbel committed:
Changed the example file to match the overall flavor

chr4 commented on init.sls:
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
 # Deploy users. Sort them, so order is not changing between salt runs
 {% for config in pillar['postgresql']['users']|default({}) %}
-createuser-{{ loop.index }}:
+createuser-{{ config['username'] }}:
```
Why did you change this? `loop.index` was used here to allow multiple rules per user.

markuskurbel replied:
I changed this to have a well defined dependency mapping in the database creation state. I am not quite sure, what you mean by "multiple roles per user", as the role is basically the user.

chr4 replied:
Sorry, I mean rules, not roles.

You need to be able to set multiple rules for a single user, like e.g. allowing two machines for replication acces:

```yaml
    - username: replicant
      type: host
      database: replication
      address: 1.2.3.4/32
      method: trust
      options: NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN REPLICATION
    - username: replicant
      type: host
      database: replication
      address: 1.2.3.5/32
      method: trust
      options: NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN REPLICATION
```

In your approach, this would result in two items with the identifier `username`, and therefore salt would complain.

markuskurbel replied:
I see. Yes this is lacking atm. This is what I meant with the initial comment, that the hba creaton needs some more care.

chr4 replied:
Well, this is actually already implemented, just don't remove the `loop.index` :)

chr4 commented on init.sls:
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
+{% for config in pillar['postgresql']['databases']|default({}) %}
 {% if config['database'] != "replication" %}
-createdb-{{ loop.index }}:
+createdb-{{ config['database'] }}:
```
I like the idea of having additional databases. Unfortunately, this would break existing configurations, were databases where created using the following pillar fields:

```yaml
users: 
  - username: test
    database: test
```

Is there a simple way to provide both syntaxes? Otherwise, a migration guide needs to be written (in README or CHANGELOG).

Another alternative would be to allow multiple databases per user, e.g. like this:

```yaml
users:
  - username: test
    databases: [bla, blubb]
```

The `database: ''` and `databases: []` pillars could even be merged automatically (not sure if this is necessary though, I suppose just supporting `databases` as an array would be sufficient).

chr4 commented on init.sls:
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
 {% endfor %}
+
+{% for config in pillar['postgresql']['schemas']|default({}) %}
+createschema-{{ config['schemaname'] }}:
```
If I understand correctly, a schema always depends on a database. Therefore, a schema could be sub-section of the database configuration?

Something like this:

```yaml
databases:
  - database: mydb
    owner: myuser
    schema: myschema
```

You are also hard-depending on a username when creating a schema, so it might even be more consise to migrate both to the user configuration section. Maybe the user section then might be a little cluttered - we should evaluate both suggestions.

markuskurbel replied:
Yes, definitely. In addition to your suggestion, a schema owner:

``` yaml
databases:
  - database: mydb
     owner: myowner
     schemas:
       - schemaname: myschema
          owner: myuser
     extensions: [foo, bar]
```

chr4 commented on init.sls:
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
-createdb-{{ loop.index }}:
+createdb-{{ config['database'] }}:
   cmd.run:
     - name: psql -t -c "CREATE DATABASE {{ config['database'] }} OWNER {{ config['username'] }};"
```
When using a seperate database configuration key, I'd uggest using the `owner` keyword instead of `username` to be more consistent with the SQL keywords.  I'm not entirely sure on this though, as we're currently using `username` instead of `role` when creating a new user as well. So this might actually be more intuitive. 
What do you think?

markuskurbel replied:
Yes, that is definitely better.

chr4 commented on init.sls:
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
+{% for config in pillar['postgresql']['schemas']|default({}) %}
+createschema-{{ config['schemaname'] }}:
+  cmd.run:
+    - name: psql -t -d {{ config['database'] }} -c "CREATE SCHEMA IF NOT EXISTS {{ config['schemaname'] }} AUTHORIZATION {{ config['username'] }};"
```
Do not use `IF NOT EXISTS`, because that would actually run the query on every salt run.
I think an `unless` clause like above would be better, because then the salt output would indicate whether a schema is actually created correctly. (This is especially useful when using `test=true`)

markuskurbel replied:
Isn't it basically the same? I mean either you get a salt log info (in the unless case) or a pg log info (in the IF NOT EXISTS case)... And either way there is SQL being executed. In case unless is False, even two queries, right?

chr4 commented on pillar.example:
```diff
@@ -66,3 +66,13 @@ postgresql:
+  databases:
+    - database: triebwerk
+      username: deploy
+      gisdb: true
```
I don't like the `gisdb` keyword here. I think a generic field `extensions` would be better, and would work with all sorts of extensions (especially built-ins)

```yaml
- database: triebwerk
  extensions: [postgis]
```

markuskurbel replied:
Yes, like the schema definition. I will change this accordingly.

chr4 commented on postgis.sls:
```diff
@@ -2,3 +2,14 @@
+  {% if config['gisdb'] %}
+postgis-extension-{{ loop.index }}:
+  cmd.run:
+    - name: psql -d {{ config['database'] }} -t -c "CREATE EXTENSION IF NOT EXISTS postgis;"
```
Please also use an `unless` guard here, same reasons as pointed out when commenting `CREATE SCHEMA`.

chr4 commented on reload.sls:
```diff
@@ -0,0 +1,9 @@
+# Reload postgres config
```
What purpose does this state have?

A database user can invoke the SQL command, and all changes should automatically be reloaded anyway when applying a state.

In case you're trying to have a state that can reload configurations on a lot of servers, I'd suggest using: `salt '*' cmd.run 'systemctl reload postgresql'` instead.

markuskurbel replied:
Whenever you create a user within a salt converge, the config needs to be reloaded. I just wanted to automate this. Your approach would work as well..I have no preference here.

chr4 replied:
This should already be done with the `watch` statement in line 11:

```yaml
    - watch:
      - file: /etc/postgresql/{{ version }}/main/*
```

markuskurbel replied:
Yes, I also thought this. However in my test runs this wasn't the case.

chr4 replied:
That's very weird, I've been using this in multiple environments and never had this problem. Can you doublecheck? The reload only gets triggert if the config file actually was changed; maybe a change was applied before and then the resulting config didn't change?

markuskurbel committed:
Fixed a typo

markuskurbel committed:
Changed some inconveniences according to PR comments:

-Renamed database user variable to owner
-Embeded the schema definition dict into the database key
-Added a extensions definition list into the database, so not just
postgis can be created

markuskurbel committed:
Changed "IF NOT EXISTS" to unless condition

markuskurbel committed:
Proposal for pg_hba.conf creation

Inspired by the saltstack-formulas/postgres-formula repo

chr4 commented on extensions.sls:
```diff
@@ -0,0 +1,16 @@
+# Install postgres extensions
+postgis:
+  pkg.installed:
+     - pkgs: [postgresql-{{ pillar['postgresql']['version'] }}-postgis-{{ pillar['postgresql']['postgis']['version'] }}]
```
Mmh, this doesn't really work - having a generic extension state, but then installing specific packages?

markuskurbel replied:
Yeah, kind of ugly. Howerver do we know that extensions are always:
`postgresql-{{ pillar['postgresql']['version'] }}-{{ extension }}-{{ pillar['postgresql']['postgis']['version'] }}` ?

chr4 commented on pillar.example:
```diff
@@ -45,24 +45,27 @@ postgresql:
+  acls:
+    - ['host', 'triebwerk', 'deploy', '10.14.38.0/24', 'trust']
+    - ['host', 'triebwerk', 'user_wwith_password', '10.14.38.0/24', 'password']
+    - ['host', 'repliction', 'replicant', '10.14.38.0/24', 'trust']
```
Mmh, why was this changed? There's no feature added and it's less readable this way and more error prone (e.g. when there's a line that doesn't have the address column)

markuskurbel replied:
Now you can have multiple rules for one user, as requested yesterday

chr4 commented on pillar.example:
```diff
@@ -45,24 +45,27 @@ postgresql:
+      schemas:
+      - schemaname: test
+        user: user_with_password
+      extensons: [postgis]
```
This looks pretty straight forward!
Is this actually a usecase to have a different schema user than the database owner?

What do you think of renaming the schema `user` attribute with `authorization` (matching the SQL `AUTHORIZATION` keyword)?

markuskurbel replied:
Yes, then I have a better control over which user can access what resource, instead of handing over the DB owner to all our users.

chr4 replied:
```yaml
  # NOTE: Database and user need to be existent
  schemas:
    - database: triebwerk
      schema: a
      authorization: postgres
```

markuskurbel commented:
After a discussion, we concluded that having the schema and extension creation under salt control is not necesssary. All other requirements are covered by the current state.
=> closed
